### DISCUSSION TITLE
To do list for distance related metric functions

### DISCUSSION THREAD

**Discussion by Khaleeh** [theochem/Selector#25] [Q&A] [answered]
@fwmeng88 my understanding is that I'm working on the distance metrics but I'm not sure where to start or what to do. should I work on the `compute_diversity` function? if possible could you provide a list of things I need to do?

**Comment by FanwangM** [comment, marked as answer]
Thank you for the question! The first actual question in our discussion board for this project.

Can you please check https://github.com/theochem/DiverseSelector/issues/4#issuecomment-1023846148 and #7 ? These two issues list several distance/similarity/dismimilairty/diversity measurements. This part is also listed in the slides that I have shared (page 9 and 10). 

You should put your codes in https://github.com/theochem/DiverseSelector/blob/main/DiverseSelector/metric.py. 

Here is a list of potential useful implementations,
1. https://github.com/VlachosGroup/AIMSim
2. https://e3fp.readthedocs.io/en/latest/api/e3fp.fingerprint.metrics.html
3. You can find more implementations online.

These resources at least can help confirm that some of your implementations are doing the right thing.

It is up to you how to set up this module, a function or a class. If you find that are too many small pieces, it's also OK to create a subfolder called *metric* with an  empty `__init__.py` file. 

Hope this explains things a little. Please feel free to keep posting questions here and tag me.

**Reply by FanwangM** [reply]
This is also helpful, `scipy.spatial.distance`, https://docs.scipy.org/doc/scipy-1.8.0/html-scipyorg/reference/spatial.distance.html#module-scipy.spatial.distance

@Khaleeh

**Comment by FanwangM** [comment]
Change log: I modified the title a little to avoid some ambiguity.

**Comment by PaulWAyers** [comment]
@fwmeng88 , looking at https://github.com/theochem/DiverseSelector/blob/1883213331c0782ba2534457a719b6598ade6680/DiverseSelector/metric.py#L41
something seems strange because it seems that the distance is "to itself" rather than a distance between two vectors?

Is there a reason to use cdist instead of pdist? I think it makes sense to use pdist where we can, as it allows constructing the whole distance matrix in a single shot, which will be useful in many cases. But maybe we should support both "computing the whole distance matrix" and not doing so, (so that kd-tree or similar can be used).

Looking at Scipy documentation, it seems like we could probably support all the scipy distance functions as one type of distance, and then support additional distance functions with a somewhat different package. If we use the same interface, when the keyword was a scipy keyword we'd default to scipy, and otherwise we'd default to our own code. (Or we could contribute to scipy, but that tends to be pretty involved.)

**Comment by FanwangM** [comment]
Thank you for the question. I will try to divide them into smaller pieces. 

### why we use `cdist` instead of `pdist`?

Suppose we have a molecule feature matrix <img src="https://render.githubusercontent.com/render/math?math=\mathbf{A} \in \mathbb{R}^{5 \times 3}"> where each column represent different molecules and each row denotes the features. What we are interested in are the pairwise distance <img src="https://render.githubusercontent.com/render/math?math=\mathbf{D} \in \mathbb{R}^{5 \times 5}"> . But if we use `pdist`, we will get a a 1D array with 10 elements, which is pairwise distance of each molecule, <img src="https://render.githubusercontent.com/render/math?math={5\choose 2}">. In order to have a distance matrix with the right shape, <img src="https://render.githubusercontent.com/render/math?math=\mathbf{D} \in \mathbb{R}^{5 \times 5}">, we will need to convert the 1D array to a symmetric 2D matrix with `pairwise_distances()` function in `scipy`, which basically filling the off-diagonal elements of a zero-matrix.

```python
# import dependencies, but will not do it again in the following blocks
import numpy as np
from scipy.spatial.distance import cdist, pdist, pairwise_distances

# mock feature matrix
a = np.array([[4, 2, 6],
                      [4, 9, 6],
                      [2, 5, 0],
                      [2, 0, 9],
                      [5, 3, 0]])
dist_pdist = pdist(a, 'euclidean')
print(dist_pdist)
# output
# [ 7.          7.          4.12310563  6.164414    7.48331477  9.69535971
  8.54400375 10.29563014  3.60555128  9.94987437]
print(dist_pdist.shape)
# output: (10,)

# convert to a symmetric matrix
dist_pdist_symm = squareform(dist_pdist)
print(dist_pdist_symm)
# output:
# [[ 0.          7.          7.          4.12310563  6.164414  ]
#  [ 7.          0.          7.48331477  9.69535971  8.54400375]
#  [ 7.          7.48331477  0.         10.29563014  3.60555128]
#  [ 4.12310563  9.69535971 10.29563014  0.          9.94987437]
#  [ 6.164414    8.54400375  3.60555128  9.94987437  0.        ]]
```

But if we use `cdist`, we can get the symmetric pairwise distance matrix directly. It is obvious that this is a better way with the same solution.
```python
dist_cdist = cdist(a, a)
print(dist_cdist)
# output
# [[ 0.          7.          7.          4.12310563  6.164414  ]
#  [ 7.          0.          7.48331477  9.69535971  8.54400375]
#  [ 7.          7.48331477  0.         10.29563014  3.60555128]
#  [ 4.12310563  9.69535971 10.29563014  0.          9.94987437]
#  [ 6.164414    8.54400375  3.60555128  9.94987437  0.        ]]
```
This is understandable. `cdist` computes distance between each pair of the two collections of inputs. Once we define <img src="https://render.githubusercontent.com/render/math?math=\mathbf{X}_{A} = a"> and <img src="https://render.githubusercontent.com/render/math?math=\mathbf{X}_{B} = a">, we are basically compute the distances between each molecule in <img src="https://render.githubusercontent.com/render/math?math=\mathbf{X}_{A}"> and each molecule in <img src="https://render.githubusercontent.com/render/math?math=\mathbf{X}_{B}">. This is exactly pairwise distance, <img src="https://render.githubusercontent.com/render/math?math={\mathbf{D} \in \mathbb{R}^{5 \times 5}}">. 

I think one catch here is that we will need a symmetric distance matrix, which will serve as the input for different selection algorithms.

### Do we have another option?

Yes, we do. The other option is the [`pairwise_distances`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances.html) in `scikit-learn` which not only supports the metrics in `scipy`, but also `[‘cityblock’, ‘cosine’, ‘euclidean’, ‘l1’, ‘l2’, ‘manhattan’]`. Originally, I was assuming that using `scipy` will not lead to a great loss of performance while `scikit-learn` is slower. Then I used `scipy`. But maybe we should try to care more about functionality and `pairwise_distances` should be the option.

### What functionalities are going to be supported in this module

The plan is to include

- pairwise distance matrix (symmetric)
- similarity matrix (symmetric)
- part of the distance matrix (we can provide the indices or mask for slciing if needed)
- conversions of the distance matrix, similarity matrix, and diversity

### Can we come up with a better way to wrap up the distance functions in `scipy` or `scikit-learn`?

Yes, of course. I will draft a wrapper soon with the support of ****kwargs** (https://book.pythontips.com/en/latest/args_and_kwargs.html and https://realpython.com/python-kwargs-and-args/), or probably with a dictionary.  For those distance matrices that are not implemented in `scipy` or `scikit-learn`, we can write our own functions. 

In the end, we may end up using a class or just a function with many private functions.

Do you have any suggestions to make it better? Any comments would be appreciated. 
@PaulWAyers @Khaleeh

**Reply by PaulWAyers** [reply]
This idea is resolved in favor of not adding dependencies now.
